core/src/main/scala/org/http4s/Compression.scala

```diff
@@ -0,0 +1,93 @@
+package org.http4s
+
+import java.util.zip.{CRC32, Deflater}
+import javax.xml.bind.DatatypeConverter
+
+import cats.Functor
+import fs2.{Chunk, Pipe, Pull, Pure, Segment, Stream}
+import fs2.Stream.chunk
+import fs2.compress.{deflate, inflate}
+import org.http4s.headers.{`Content-Encoding`, `Content-Length`, `Transfer-Encoding`}
+import org.log4s.getLogger
+
+object Compression {
+
+  private[this] val logger = getLogger
+
+  def zipResponse[F[_]: Functor](bufferSize: Int, level: Int, resp: Response[F]): Response[F] = {
```

Comment:
I think this could take a `Message[F]` instead of a `Response[F]`.  Couldn't this apply to requests as well as responses?

Reply:
If we go for `Message[F]` then we have to work with the function `withBody` instead of `copy` which requires a `Monad` for `F`.